File: platforms/ios/Dezel/Dezel/Core/JavaScriptFinalizeCallback.swift
/**
 * @class JavaScriptFinalizeCallback
 * @since 0.1.0
 */
public final class JavaScriptFinalizeCallback {

	//--------------------------------------------------------------------------
	// MARK: Properties
	//--------------------------------------------------------------------------

    /**
     * @property context
     * @since 0.1.0
	 * @hidden
     */
	private(set) public var context: JavaScriptContext

    /**
     * @property handle
     * @since 0.4.0
	 * @hidden
     */
	private(set) public var handle: JavaScriptValueDataRef

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

    /**
     * @constructor
     * @since 0.1.0
	 * @hidden
     */
	internal init(context: JavaScriptContext, handle: JavaScriptValueDataRef) {
		self.context = context
		self.handle = handle
	}

	/**
	 * @method attribute
	 * @since 0.4.0
	 */
	public func attribute(_ key: AnyObject) -> AnyObject? {
		return toUnretainedObject(JavaScriptValueDataGetAttribute(self.handle, toHash(key)))
	}

	/**
	 * @method attribute
	 * @since 0.4.0
	 */
	public func attribute(_ key: AnyObject, value: AnyObject?) {
		let hash = toHash(key)
		JavaScriptValueDataGetAttribute(self.handle, hash)?.release()
		JavaScriptValueDataSetAttribute(self.handle, hash, toRetainedOpaque(value))
	}
}

File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/core/JavaScriptFinalizeCallback.kt
package ca.logaritm.dezel.core

import ca.logaritm.dezel.core.external.JavaScriptValueExternal

/**
 * @class JavaScriptFinalizeCallback
 * @since 0.1.0
 */
public class JavaScriptFinalizeCallback(context: JavaScriptContext, handle: Long) {

	//--------------------------------------------------------------------------
	// Properties
	//--------------------------------------------------------------------------

	/**
	 * @property context
	 * @since 0.1.0
	 */
	public var context: JavaScriptContext
		private set

	/**
	 * @property handle
	 * @since 0.1.0
	 */
	public var handle: Long = 0
		private set

	//--------------------------------------------------------------------------
	// Method
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.1.0
	 * @hidden
	 */
	init {
		this.context = context
		this.handle = handle
	}

	/**
	 * @method attribute
	 * @since 0.4.0
	 */
	public fun attribute(key: Any): Any? {
		return JavaScriptValueExternal.getAttribute(this.handle, key.hashCode())
	}

	/**
	 * @method attribute
	 * @since 0.4.0
	 */
	public fun attribute(key: Any, value: Any?) {
		val hash = key.hashCode()
		JavaScriptValueExternal.delAttribute(this.handle, hash)
		JavaScriptValueExternal.setAttribute(this.handle, hash, value)
	}
}